### PR TITLE
Python AlphaZero

### PR DESCRIPTION
**Pull request by sbodenstein**
This AlphaZero implementation has many design issues that need to be resolved before finalizing the PR (eg. there are no tests, and input validation is not production-ready yet). It trains well on tic-tac-toe (see `examples/tic_tac_toe_alpha_zero.py`), giving confidence that the implementation is correct.

The philosophy of this implementation is to follow the pseudocode of the AlphaZero [Science paper](https://science.sciencemag.org/content/362/6419/1140) as closely as possible, and also fit it into the current OpenSpiel framework with minimal framework changes. But here are some differences with the pseudocode:
- It uses very large batch sizes (`>4000`) and uses gradient accumulation to allow this to fit into memory. However, the AlphaZero ResNet architecture uses batchnorm, which doesn't play well with gradient accumulation. So I've avoided this
- I've used ADAM as the default optimizer, whilst the pseudocode uses SGD with momentum and effectively a `tf.train.piecewise_constant` learning rate scheduler. I've found this very annoying to use for adapting to new games as there are now some extra fiddly hyperparameters that need tweaking.
- The loss was not scaled by the batch size in the pseudocode. This makes it harder to change the batch size as other hyperparameters are now coupled to it (eg learning rate, l2 regularization). So this implementation does normalize by batch size.
- It doesn't support non-vector actions representations (eg what they did for Chess). But the paper does also mention that `We also tried using a flat distribution over moves for chess and shogi; the final result was almost identical although training was slightly slower`. So left this feature out for now.
- Missing support for `max_moves`: this seems like a feature to add to the Game class, but it currently seems hardcoded?
- No clean way of handling almost-Markovian environments (eg Chess and Go) by training on some context window of previous observations. This might be best handled at the game-level? (or am I missing some library feature?)

Some design explanations and issues:
- The main framework-level change is to introduce a `TrainableEvaluator` class that acts like the `Evaluator` class, but also has an `update` method with training logic (doesn't have to be neural net training). It also builds caching into the design due to how common using nets with both value and policy heads is (so subclasses would only implement the `value_and_prior` method).
- More on caching in `mcts.TrainableEvaluator`: The current design of `mcts.Evaluator` is that `evaluate` and `prior` are called separately during mcts search. In the case where these are (expensively) computed jointly as in AlphaZero, this is inefficient. I am well-aware of the general injunction that `OpenSpiel provides reference implementations that are used to learn from and prototype with, rather than fully-optimized / high-performance code`. So I am aware that this caching mechanism might be contentious, but self-play is the bottleneck and makes AlphaZero a lot less pleasant to use.
- The AlphaZero training loop: this implementation ties the number of training iterations to the current buffer size (via `num_training_epochs` in `AlphaZero.update`). This has some virtues: the epoch number to train over is a much more game-independent hyperparameter, and makes more sense than the AlphaZero pseudocode approach (do a huge number of training iterations even when the buffer hasn't been filled up yet).  One option I considered first is to make `AlphaZero.update` do a single training iteration, which gives most flexibility to users. But: the optimizer needs to be reset after all training iterations are done. The user could call a reset method, which is ugly and easy to forget to do. Otherwise, the optimizer reset could be done in `AlphaZero.self_play`. But this side effect of `AlphaZero.self_play` also feels a bit hacky. I'm definitely not confident I've chosen the right design here.
- Device flags for NN: I didn't see any other examples of GPU support in OpenSpiel. So went ahead and allowed the strings `'gpu'`, `'cpu'`, or a `tf.device(...)` object (useful to specify more exotic things, like a particular GPU on a multi-GPU machine. Or maybe some other device type like TPU). 
- Neural nets. In all other algorithms that use neural nets, there is a lack of flexibility in specifying architectures. This could well be intentional (makes it easier to write backend independent implementations as there is a nice backend agnostic interface with only knobs like `hidden_layers_sizes` that can be tweaked). I've gone against this, and allowed any Keras model to be supplied to `AlphaZeroKerasEvaluator`, along with allowing flexible feature extraction. I've also focused on Keras, as its official high-level TF 2.0 API (I assume OpenSpiel will move to TF 2.0 at some stage).
- AlphaZero can probably be extended to more than 2 agents (and the MCTS implementation already supports that). But that can be a future extension, as this implementation only supports 2-player games.

### CODE REVIEW

**Comment by lanctot** [comment]
Wow, @sbodenstein, thanks for this. I will unfortunately be away on vacation for two weeks and then at AAAI, so will have to take a look and follow-up when I'm back.

**Comment by sbodenstein** [comment]
@lanctot: no problem at all, enjoy your holiday!

**Comment by tewalds** [comment]
@sbodenstein Could you sign the CLA?

**Comment by tewalds** [comment]
Btw, I don't mind fixing up the PR for inclusion, and certainly don't ask that you add the parallelism or model/evaluator refactoring for inclusion, but I can't import it at all without the CLA being signed.

**Comment by sbodenstein** [comment]
@tewalds: Apologies about that, I did sign the CLA, but there were two commits by @taliesinb, and he signed the CLA but didn't write the comment for `googlebot`. He will do this later today.

File: open_spiel/python/algorithms/masked_softmax.py [hunk 1]
```diff
@@ -54,7 +54,8 @@ def np_masked_softmax(logits, legal_actions_mask):
     legal_actions_mask: The legal action mask, same shape as logits. 1 means
       it's a legal action, 0 means it's illegal.
   """
-  masked_logits = logits + np.log(legal_actions_mask)
+  with np.errstate(divide='ignore'):
```
[the remark this answers is not in the document]
Comment (sbodenstein) [reply open_spiel/python/algorithms/masked_softmax.py]: Calling `np_masked_softmax` raises division by zero warning (`RuntimeWarning: divide by zero encountered in log`), which pollutes the terminal when its called often. This suppresses the warning.

**Comment by taliesinb** [comment]
@googlebot googlebot I fixed it..

**Comment by taliesinb** [comment]
@googlebot I consent.

File: open_spiel/python/algorithms/dqn.py [hunk 2]
```diff
@@ -62,7 +62,7 @@ def add(self, element):
       self._next_entry_index += 1
       self._next_entry_index %= self._replay_buffer_capacity
 
-  def sample(self, num_samples):
```
[the remark this answers is not in the document]
Comment (sbodenstein) [reply open_spiel/python/algorithms/dqn.py]: Happy to do that. One alternative is to inherit from `dqn.ReplayBuffer` like what is done in `algorithms/eva.py`, and then override the `sample` method. But agreed that long-term, makes sense to have buffer library.

File: open_spiel/python/algorithms/alpha_zero.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+      l2_loss += l2
+
+    loss_str = "Epoch {0} mean losses. Total: {1:.3g}, Policy: {2:.3g}, Value: {3:.3g}, L2: {4:.3g}"
+    n = float(len(losses))
+    print(
+        loss_str.format(epoch, total_loss / n, policy_loss / n, value_loss / n,
+                        l2_loss / n))
+
+
+def alpha_zero_ucb_score(child, parent_explore_count, params):
```
[the remark this answers is not in the document]
Comment (sbodenstein) [reply open_spiel/python/algorithms/alpha_zero.py]: This was used for Chess and Go in the AlphaZero paper, but I found it worked far less well for tic-tac-toe. I originally had this as the default UCB score method, but changed it due to the tic-tac-toe experiments. Should probably switch back to using this as the default, as it was used in the paper/pseudocode.

[the remark this answers is not in the document]
Comment (tewalds) [reply open_spiel/python/algorithms/alpha_zero.py]: MCTSBot already supports uct and puct.

File: open_spiel/python/algorithms/dqn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tewalds) [reply open_spiel/python/algorithms/dqn.py]: It'd be good to start with it self-contained. I can do that as well when importing this.

**Comment by sbodenstein** [comment]
@tewalds: regarding your general comments:

> it'd be nice to have a separation between the model and the evaluator so the learn step is part of the model, not the evaluator.

Happy to do that. Regarding the class structure, here is a proposal: the `__init__` and `update` methods would look the same as the current `AlphaZeroKerasEvaluator`. It would have a `evaluate` or `forward` method that returns a tuple of outputs (one for each head), and the class is callable (calling `evaluate` or `forward`). It might also have a `save` method for saving the net to disk.

It would also make sense for the caching to happen at the model-level now. 

> it'd be nice to have some parallelism so this can train on bigger games

This would be very nice indeed. Was not sure how far the injunction `OpenSpiel provides reference implementations that are used to learn from and prototype with, rather than fully-optimized / high-performance code` went 😉

Do you have a parallelization method in mind?

File: open_spiel/python/algorithms/alpha_zero.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbodenstein) [reply open_spiel/python/algorithms/alpha_zero.py]: It does, but not the AlphaZero variant (unless I'm missing something). For example, the AlphaZero variant has two hyperparameters, rather than just the one in `mcts.py`. From the pseudocode,
```
def ucb_score(config: AlphaZeroConfig, parent: Node, child: Node):
  pb_c = math.log((parent.visit_count + config.pb_c_base + 1) /
                  config.pb_c_base) + config.pb_c_init
  pb_c *= math.sqrt(parent.visit_count) / (child.visit_count + 1)

  prior_score = pb_c * child.prior
  value_score = child.value()
  return prior_score + value_score
```
where `pb_c_base = 19652` and `pb_c_init = 1.25` are the defaults.

**Comment by lanctot** [comment]
> This would be very nice indeed. Was not sure how far the injunction OpenSpiel provides reference implementations that are used to learn from and prototype with, rather than fully-optimized / high-performance code went 😉

Ha, that is all me :) The main reason I stated this was to make sure all the implementations are readable and can be understood by newbies. I have read many minimax search implementations that, e.g. use bitboard transposition tables and heuristics++, that would not fall into thos category. So what I am trying to prevent are things like premature or unnecessary optimization.

However, if an optimization is necessary to do research, and if we can do it in a way that is still general and friendly to newcomers, then it's probably worth doing IMO.

**Comment by lanctot** [comment]
And my main reason for this is that I want OpenSpiel to be friendly to newcomers above all else. There are of course always trade-offs, but if we can preserve this property, then we can appeal to a much larger crowd, which I think is a long-term win. Just my $0.02 on this subject. I appreciate bringing it up, very important point IMO.
